compat/bionic/pthread_cancel.h

```diff
@@ -0,0 +1,66 @@
+/*
+* pthread_cancel compatibility implementation for Android
+* Copyright (c) 2015 Paul Idstein
+*
+* This file is part of FFmpeg.
+*
+* FFmpeg is free software; you can redistribute it and/or
+* modify it under the terms of the GNU Lesser General Public
+* License as published by the Free Software Foundation; either
+* version 2.1 of the License, or (at your option) any later version.
+*
+* FFmpeg is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public
+* License along with FFmpeg; if not, write to the Free Software
+* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+*/
+#ifndef _PTHREAD_CANCEL_H_
+#define _PTHREAD_CANCEL_H_
+/* Alternative cancel implementation that is used on non POSIX compliant
+platforms such as Android.
+We will send a user-defined signal to a thread, then check for
+thread_canceled to trigger thread teardown.
+
+Attention: It's the user's duty to implement a proper
+clean-up handler! */
+
+#include <pthread.h>
+
+/* Signals are on a per-process base.
+SIGUSR1 is already in use by video grab interface*/
+#define SIG_CANCEL_SIGNAL SIGUSR2
+#define PTHREAD_CANCEL_ENABLE 1
+#define PTHREAD_CANCEL_DISABLE 0
+static __inline__ int pthread_setcancelstate(int state, int *oldstate) {
+  sigset_t   new, old;
+  int ret;
+  sigemptyset(&new);
+  sigaddset (&new, SIG_CANCEL_SIGNAL);
+  /* We block CANCEL signal to see it as pending
+     or unblock to let it be discarded */
+  ret = pthread_sigmask(state == PTHREAD_CANCEL_ENABLE ? SIG_BLOCK : SIG_UNBLOCK, &new , &old);
+  if(oldstate != NULL)
```

Comment:
It may happen that the cancel state is changed between two AS-safe method invocations. The behaviour then is unpredictable. Make sure to end up in a consistent state by ensuring only a very small amount of logic between AS-safe calls. Otherwise the incoming signals are not clearly verified!